src/mscorlib/src/System/Text/ASCIIEncoding.cs

```diff
@@ -78,12 +80,158 @@ public override unsafe int GetByteCount(char* chars, int count)
+                {
+                    // Definitely not enough space, early bail
+                    EncodingForwarder.ThrowBytesOverflow(this);
+                }
+                fixed (char* pInput = chars)
+                fixed (byte* pOutput = &bytes[0]) 
+                {
+                    char* input = pInput + charIndex;
+                    byte* output = pOutput + byteIndex;
+                    int charactersConsumed;
+                    if (!EncodingForwarder.TryEncode(input, charCount, output, byteCount, out charactersConsumed, out bytesWritten)) 
+                    {
+                        // Not all ASCII, GetBytesFallback for remaining conversion
+                        bytesWritten += GetBytesFallback(input + charactersConsumed, charCount - charactersConsumed, output + bytesWritten, byteCount - bytesWritten, null);
+                    }
+                }
+            } 
+            else 
+            {
+                // Nothing to encode
+                bytesWritten = 0;
+            }
+
+            return bytesWritten;
+        }
+
+        public override byte[] GetBytes(char[] chars) {
+            if (chars == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.chars, ExceptionResource.ArgumentNull_Array);
+            Contract.EndContractBlock();
+            
+            return GetBytesValidated(chars, 0, chars.Length);
+        }
+
+        public override byte[] GetBytes(char[] chars, int index, int count) {
+            if ((chars == null) ||
+                (index < 0) ||
+                (count < 0) ||
+                (chars.Length - index < count))
+            {
+                EncodingForwarder.ThrowValidationFailedException(chars, index, count);
+            }
+            Contract.EndContractBlock();
+            
+            return GetBytesValidated(chars, index, count);
+        }
+
+        private unsafe byte[] GetBytesValidated(char[] chars, int index, int count)
+        {
+            byte[] bytes;
+            if (count > 0)
+            {
+                fixed (char* input = chars)
+                {
+                    bytes = GetBytesValidated(input + index, count);
+                }
+            }
+            else
+            {
+                bytes = Array.Empty<byte>();
+            }
+
+            return bytes;
+        }
+
+        private unsafe byte[] GetBytesValidated(char* input, int charCount)
         {
-            return EncodingForwarder.GetBytes(this, chars, charIndex, charCount, bytes, byteIndex);
+            int remaining = 0;
+            // Assume string is all ASCII and size array for that
+            byte[] bytes = new byte[charCount];
+
+            int bytesWritten;
+            fixed (byte* output = &bytes[0]) 
+            {
+                int charactersConsumed;
+                if (!EncodingForwarder.TryEncode(input, charCount, output, charCount, out charactersConsumed, out bytesWritten)) 
```

Comment:
>charactersConsumed, out bytesWritten [](start = 90, length = 36)

do you think we'll have a case that charactersConsumed will be different than byteswritten?